File: frameworks/actix-web-rest/tests/mutations_test.rs
mod test_helpers;
use test_helpers::*;

// ============================================================================
// Mutation Simulation: updateUser
// ============================================================================

#[test]
fn test_update_user_full_name() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "Developer");
    let user_id = user.id.clone();

    // Simulate mutation
    user.full_name = "Alice Smith".to_string();

    // Verify
    assert_eq!(user.full_name, "Alice Smith");
    assert_eq!(user.id, user_id);
}

#[test]
fn test_update_user_bio() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "Developer");
    let user_id = user.id.clone();

    // Simulate mutation
    user.bio = Some("Senior Developer".to_string());

    // Verify
    assert_eq!(user.bio, Some("Senior Developer".to_string()));
    assert_eq!(user.id, user_id);
}

#[test]
fn test_update_user_both_fields() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "Developer");
    let user_id = user.id.clone();

    // Simulate mutation
    user.full_name = "Alice Smith".to_string();
    user.bio = Some("Senior Developer".to_string());

    // Verify
    assert_eq!(user.full_name, "Alice Smith");
    assert_eq!(user.bio, Some("Senior Developer".to_string()));
    assert_eq!(user.id, user_id);
}

#[test]
fn test_update_user_clear_bio() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "Developer");
    let user_id = user.id.clone();

    // Simulate mutation
    user.bio = None;

    // Verify
    assert_eq!(user.bio, None);
    assert_eq!(user.id, user_id);
}

// ============================================================================
// Mutation Simulation: updatePost
// ============================================================================

#[test]
fn test_update_post_title() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let mut post = factory.create_test_post(&author.id, "Original Title", "Original Content");
    let post_id = post.id.clone();

    // Simulate mutation
    post.title = "Updated Title".to_string();

    // Verify
    assert_eq!(post.title, "Updated Title");
    assert_eq!(post.id, post_id);
}

#[test]
fn test_update_post_content() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let mut post = factory.create_test_post(&author.id, "Original Title", "Original Content");
    let post_id = post.id.clone();

    // Simulate mutation
    post.content = Some("Updated Content".to_string());

    // Verify
    assert_eq!(post.content, Some("Updated Content".to_string()));
    assert_eq!(post.id, post_id);
}

#[test]
fn test_update_post_both_fields() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let mut post = factory.create_test_post(&author.id, "Original Title", "Original Content");
    let post_id = post.id.clone();

    // Simulate mutation
    post.title = "Updated Title".to_string();
    post.content = Some("Updated Content".to_string());

    // Verify
    assert_eq!(post.title, "Updated Title");
    assert_eq!(post.content, Some("Updated Content".to_string()));
    assert_eq!(post.id, post_id);
}

// ============================================================================
// Field Immutability
// ============================================================================

#[test]
fn test_user_id_immutable_after_update() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "Bio");
    let original_id = user.id.clone();

    // Try to "update"
    user.bio = Some("Updated".to_string());

    // Verify ID unchanged
    assert_eq!(user.id, original_id);
}

#[test]
fn test_post_id_immutable_after_update() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let mut post = factory.create_test_post(&author.id, "Title", "Content");
    let original_id = post.id.clone();

    // Try to "update"
    post.title = "Updated".to_string();

    // Verify ID unchanged
    assert_eq!(post.id, original_id);
}

#[test]
fn test_username_immutable() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    let original_username = user.username.clone();

    // Try to "update"
    user.bio = Some("Updated".to_string());

    // Verify username unchanged
    assert_eq!(user.username, original_username);
}

// ============================================================================
// State Changes
// ============================================================================

#[test]
fn test_sequential_updates_accumulate() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "");

    // Apply updates sequentially
    user.bio = Some("Developer".to_string());
    user.bio = Some("Senior Developer".to_string());

    // Verify latest state
    assert_eq!(user.bio, Some("Senior Developer".to_string()));
}

#[test]
fn test_updates_isolated_between_entities() {
    let factory = TestFactory::new();
    let mut user1 = factory.create_test_user("alice", "alice@example.com", "Alice", "Bio1");
    let mut user2 = factory.create_test_user("bob", "bob@example.com", "Bob", "Bio2");

    let original_bio2 = user2.bio.clone();

    // Update user1
    user1.bio = Some("Updated".to_string());

    // Verify user2 unchanged
    assert_eq!(user2.bio, original_bio2);
}

// ============================================================================
// Return Value Validation
// ============================================================================

#[test]
fn test_updated_user_returns_all_fields() {
    let factory = TestFactory::new();
    let mut user = factory.create_test_user("alice", "alice@example.com", "Alice", "Developer");
    user.bio = Some("Updated".to_string());

    // Verify all fields present
    assert!(!user.id.is_empty());
    assert_eq!(user.username, "alice");
    assert_eq!(user.bio, Some("Updated".to_string()));
}

#[test]
fn test_updated_post_returns_all_fields() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let mut post = factory.create_test_post(&author.id, "Title", "Content");
    post.title = "Updated".to_string();

    // Verify all fields present
    assert!(!post.id.is_empty());
    assert_eq!(post.title, "Updated");
    assert_eq!(post.author_id, author.id);
}

File: frameworks/java-spring-boot/src/test/java/com/fraiseql/MutationsTest.java
package com.fraiseql;

import com.fraiseql.models.User;
import com.fraiseql.models.Post;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class MutationsTest {
    private TestFactory factory;

    @BeforeEach
    void setUp() {
        factory = new TestFactory();
    }

    // ============================================================================
    // Mutation: updateUser
    // ============================================================================

    @Test
    void testUpdateUserFullName() {
        User user = factory.createTestUser("alice", "alice@example.com", "Alice", "Developer");
        String userId = user.getId();

        // Simulate mutation
        user.setFullName("Alice Smith");

        // Verify
        assertEquals("Alice Smith", user.getFullName());
        assertEquals(userId, user.getId());
    }

    @Test
    void testUpdateUserBio() {
        User user = factory.createTestUser("alice", "alice@example.com", "Alice", "Developer");
        String userId = user.getId();

        // Simulate mutation
        user.setBio("Senior Developer");

        // Verify
        assertEquals("Senior Developer", user.getBio());
        assertEquals(userId, user.getId());
    }

    @Test
    void testUpdateUserBothFields() {
        User user = factory.createTestUser("alice", "alice@example.com", "Alice", "Developer");
        String userId = user.getId();

        // Simulate mutation
        user.setFullName("Alice Smith");
        user.setBio("Senior Developer");

        // Verify
        assertEquals("Alice Smith", user.getFullName());
        assertEquals("Senior Developer", user.getBio());
        assertEquals(userId, user.getId());
    }

    @Test
    void testUpdateUserClearBio() {
        User user = factory.createTestUser("alice", "alice@example.com", "Alice", "Developer");
        String userId = user.getId();

        // Simulate mutation
        user.setBio(null);

        // Verify
        assertNull(user.getBio());
        assertEquals(userId, user.getId());
    }

    // ============================================================================
    // Mutation: updatePost
    // ============================================================================

    @Test
    void testUpdatePostTitle() {
        User author = factory.createTestUser("author", "author@example.com", "Author", "");
        Post post = factory.createTestPost(author.getId(), "Original Title", "Original Content");
        String postId = post.getId();

        // Simulate mutation
        post.setTitle("Updated Title");

        // Verify
        assertEquals("Updated Title", post.getTitle());
        assertEquals(postId, post.getId());
    }

    @Test
    void testUpdatePostContent() {
        User author = factory.createTestUser("author", "author@example.com", "Author", "");
        Post post = factory.createTestPost(author.getId(), "Original Title", "Original Content");
        String postId = post.getId();

        // Simulate mutation
        post.setContent("Updated Content");

        // Verify
        assertEquals("Updated Content", post.getContent());
        assertEquals(postId, post.getId());
    }

    @Test
    void testUpdatePostBothFields() {
        User author = factory.createTestUser("author", "author@example.com", "Author", "");
        Post post = factory.createTestPost(author.getId(), "Original Title", "Original Content");
        String postId = post.getId();

        // Simulate mutation
        post.setTitle("Updated Title");
        post.setContent("Updated Content");

        // Verify
        assertEquals("Updated Title", post.getTitle());
        assertEquals("Updated Content", post.getContent());
        assertEquals(postId, post.getId());
    }

    // ============================================================================
    // Field Immutability
    // ============================================================================

    @Test
    void testUserIdImmutableAfterUpdate() {
        User user = factory.createTestUser("alice", "alice@example.com", "Alice", "Bio");
        String originalId = user.getId();

        // Try to "update"
        user.setBio("Updated");

        // Verify ID unchanged
        assertEquals(originalId, user.getId());
    }

    @Test
    void testPostIdImmutableAfterUpdate() {
        User author = factory.createTestUser("author", "author@example.com", "Author", "");
        Post post = factory.createTestPost(author.getId(), "Title", "Content");
        String originalId = post.getId();

        // Try to "update"
        post.setTitle("Updated");

        // Verify ID unchanged
        assertEquals(originalId, post.getId());
    }

    @Test
    void testUsernameImmutable() {
        User user = factory.createTestUser("alice", "alice@example.com", "Alice", "");
        String originalUsername = user.getUsername();

        // Try to "update"
        user.setBio("Updated");

        // Verify username unchanged
        assertEquals(originalUsername, user.getUsername());
    }

    // ============================================================================
    // State Changes
    // ============================================================================

    @Test
    void testSequentialUpdatesAccumulate() {
        User user = factory.createTestUser("alice", "alice@example.com", "Alice", "");

        // Apply updates sequentially
        user.setBio("Developer");
        user.setBio("Senior Developer");

        // Verify latest state
        assertEquals("Senior Developer", user.getBio());
    }

    @Test
    void testUpdatesIsolatedBetweenEntities() {
        User user1 = factory.createTestUser("alice", "alice@example.com", "Alice", "Bio1");
        User user2 = factory.createTestUser("bob", "bob@example.com", "Bob", "Bio2");

        String originalBio2 = user2.getBio();

        // Update user1
        user1.setBio("Updated");

        // Verify user2 unchanged
        assertEquals(originalBio2, user2.getBio());
    }

    // ============================================================================
    // Return Value Validation
    // ============================================================================

    @Test
    void testUpdatedUserReturnsAllFields() {
        User user = factory.createTestUser("alice", "alice@example.com", "Alice", "Developer");
        user.setBio("Updated");

        // Verify all fields present
        assertNotNull(user.getId());
        assertEquals("alice", user.getUsername());
        assertEquals("Updated", user.getBio());
    }

    @Test
    void testUpdatedPostReturnsAllFields() {
        User author = factory.createTestUser("author", "author@example.com", "Author", "");
        Post post = factory.createTestPost(author.getId(), "Title", "Content");
        post.setTitle("Updated");

        // Verify all fields present
        assertNotNull(post.getId());
        assertEquals("Updated", post.getTitle());
        assertNotNull(post.getAuthor());
    }

    @Test
    void testMutationMaintainsCreatedAt() {
        User user = factory.createTestUser("alice", "alice@example.com", "Alice", "");
        var originalCreatedAt = user.getCreatedAt();

        // Update
        user.setFullName("Alice Updated");

        // Verify created_at unchanged
        assertEquals(originalCreatedAt, user.getCreatedAt());
    }
}
